### ISSUE TITLE
Evaluation of ms lesion segmentation algorithm

### ISSUE THREAD

**Issue by plbenveniste** [ivadomed/ms-lesion-agnostic#21]
Opening this issue to discuss the results and the performances of the models train to segment MS lesions in the spinal cord.

**Comment by plbenveniste** [comment]
Discussing the results of the model output stored on `ms_lesion_agnostic/results/2024-04-21_16/06/04.890513`

Here is the model performance on the training set (data from canproco and sct-testing-large)
![dice_scores](https://github.com/ivadomed/ms-lesion-agnostic/assets/67429280/40be8b87-c3c3-4be4-918b-adf9af3c7914)

Here are the detailed observations:

- Lesions GT segmentations that should be corrected
    - sub-edm181_ses-M0_PSIR
    - sub-cal156_ses-M0_STIR
    - sub-cal104_ses-M0_STIR: lesions are too small to be kept
    - sub-mon004_ses-M0_PSIR
    - sub-mon010_ses-M12_PSIR (lesions are too small): either remove them on the lesion segmentation or remove during pre-processing
    - sub-mon137_ses-M0_PSIR (to be discussed)
    - sub-karo1898_acq-sagcerv_T2star (to be discussed)
    - sub-karo2032_acq-sagcerv_T2star
    - sub-karo2039_acq-sagcerv_T2star
    - sub-nyuShepherd022_acq-sup_T2w
- Identified problems:
  - The model tends to segment lesions which are too small. Problem: some input data contains lesion segmentations which are very small. Solution: either remove those lesions on the lesion segmentation directly or remove them during pre-processing or remove them after model prediction?
  - The model doesn’t segment lesions close to the brain. (ex: sub-amuVirginie009_T2w, sub-bwh026_acq-sagstir_T2w, sub-lyonOfsep004_acq-sag_T2w, sub-lyonOfsep082_acq-sag_T2w, sub-milanFilippi064_acq-sag_T2w)
  - In larger fields of view than canproco, model usually doesn't segment lesions close to the brain, but still segments lesions in the cervical/thoracic spinal cord: (sub-lyonOfsep001_acq-sag_T2w)
  - The subject sub-lyonOfsep079_acq-sag_T2w : should be excluded as problem when reconstructing the image
  - Rare issue of model segmenting outside the spinal cord : sub-rennesMS018_acq-sagthor_T2w, sub-rennesMS027_acq-sagthor_T2w, sub-rennesMS049_acq-sagthor_T2w: maybe because spinal cord is not centered. The model tends to segment in the middle (on the vertical axis)

Conclusion: 
- Because of the use of `RandCropByPosNegLabeld`, the model tends to segment in the middle of the spinal cord (on the superior-inferior axis): this can be solved by modifying/removing this DA strategy.
- Lesions which are below a certain volume (to be defined) should not be kept during model training and during inference.

**Comment by plbenveniste** [comment]
<details>

<summary>Below the GIFs of the problematic images quoted above: </summary>

`sub-edm181_ses-M0_PSIR` 
![ezgif com-animated-gif-maker (4)](https://github.com/ivadomed/ms-lesion-agnostic/assets/67429280/0c8e320e-30a5-439c-b102-ba624f2561b2)

`sub-cal156_ses-M0_STIR` 
![ezgif com-animated-gif-maker (5)](https://github.com/ivadomed/ms-lesion-agnostic/assets/67429280/c693ab3c-744b-4858-bcf1-65620bf49404)

`sub-cal104_ses-M0_STIR`
![ezgif com-animated-gif-maker (6)](https://github.com/ivadomed/ms-lesion-agnostic/assets/67429280/38a228ba-1751-466e-b3ce-037440dcca65)

`sub-mon004_ses-M0_PSIR`
![ezgif com-animated-gif-maker (7)](https://github.com/ivadomed/ms-lesion-agnostic/assets/67429280/2f14a106-b932-49b8-8fb6-fd13c1298413)

`sub-mon010_ses-M12_PSIR`
![ezgif com-animated-gif-maker (8)](https://github.com/ivadomed/ms-lesion-agnostic/assets/67429280/d6752291-b551-4d34-95c1-9f1d25d70b1b)

`sub-mon137_ses-M0_PSIR`
![ezgif com-animated-gif-maker (9)](https://github.com/ivadomed/ms-lesion-agnostic/assets/67429280/7dc35dcb-ef19-41a0-8c07-3cf0bdff1007)

`sub-karo1898_acq-sagcerv_T2star`
![ezgif com-animated-gif-maker (10)](https://github.com/ivadomed/ms-lesion-agnostic/assets/67429280/f7cf357f-37b7-411e-abcc-3382155fc85b)

`sub-karo2039_acq-sagcerv_T2star`
![ezgif com-animated-gif-maker (11)](https://github.com/ivadomed/ms-lesion-agnostic/assets/67429280/19cf03e2-b700-43d0-a9c6-2b1ce0272ed7)

`sub-nyuShepherd022_acq-sup_T2w`
![ezgif com-animated-gif-maker (12)](https://github.com/ivadomed/ms-lesion-agnostic/assets/67429280/5a42a93d-27ca-433d-aa93-98885e52c2ce)


</details>

**Comment by plbenveniste** [comment]
Here is a graph presenting the results of the investigations done: 
![image](https://github.com/ivadomed/ms-lesion-agnostic/assets/67429280/9fc47577-8817-4756-9163-a6f1e418e1f2)

This shows a comparison between: 
- the current state of the art model on which the above observations were done (called BestAttention+spacing1;0)
- the SOTA model without the RandCropByPosNeg augmentation function
- the SOTA model without the RandCropByPosNeg augmentation function and no skipping of empty patch
- the SOTA model with a SpatialCrop function on top of RandCropByPosNeg to add variability
- the SOTA model with the removeSmall augmentation function

As we can see on the graphs, the explorations did not outperform the current SOTA model. 

As suggested by @jcohenadad exploring the nnUNet data augmentation strategies could help : [link to data augmentation](https://github.com/MIC-DKFZ/nnUNet/blob/75c46fe7c1c85c60e020d76e387772c7be02a1af/nnunetv2/training/nnUNetTrainer/nnUNetTrainer.py#L14C1-L38C3)

**Comment by jcohenadad** [comment]
Very interesting investigations @plbenveniste !

**Comment by plbenveniste** [comment]
The current investigation is to train a MedNext model for MS lesion segmentation:
- The first investigation was to train a MedNext model with a low number of channels : 3 (to see if it works). It had a weird behavior as we can see on the following graph (comparison with the SOTA model in violet and the low channel MedNext model). 
<img width="662" alt="image" src="https://github.com/ivadomed/ms-lesion-agnostic/assets/67429280/3e3ac34c-d8e1-4e1f-9828-bd4052d3ef97">
- The current MedNext model is with the highest number of channel possible (n_channels=16) due to the GPU limitations (using 42GB out of 47GB). It is currently training, but it seems that it is slightly under-performing the SOTA model. 
- The next investigation I want to do is to increase the number of channel in MedNext and, to do so, reduce the number of samples produced by RandCropByPosNeg (num_samples=4).

**Comment by plbenveniste** [comment]
Other findings:
- using DiceFocalLoss made training fail
- using GeneralizedDiceFocalLoss made training fail
- using a model depth which ranged from 64 to 1024 improved slightly (~2%) the model performances: however, it requires a lot more memory. This made not possible training with RandCropByPosNeg with a num_sample greater than 2. 

For now, my conclusions are: 
- the best resampling resolution is 0.7 isotropic
- the nnUNet data augmentation didn't improve the model performance
- the best model architecture is AttUnet with depth [64, 128, 256, 512, 1024]

TODO: 

- [ ] Use the best model to perform inference on the training and validation images. 
- [ ] Identify the images on which the model is poorly performing
- [ ] Manually correct these images and/or remove than from the training/validation (using an exclude.yml file for instance)
- [ ] Retrain the model

**Comment by plbenveniste** [comment]
The best model is stored in `/home/plbenveniste/net/ms-lesion-agnostic/results/2024-07-18_10:46:21.634514/` it is called `
BestAtt+allData_06-26` in wandb. 

The inference was performed on the trainining, validation and testing set using the following command on kronos: 
```console
conda activate venv_monai
CUDA_VISIBLE_DEVICES=1 python ms-lesion-agnostic/monai/test_model.py --config ms-lesion-agnostic/monai/config_test.yml --data-split train
CUDA_VISIBLE_DEVICES=2 python ms-lesion-agnostic/monai/test_model.py --config ms-lesion-agnostic/monai/config_test.yml --data-split test
CUDA_VISIBLE_DEVICES=3 python ms-lesion-agnostic/monai/test_model.py --config ms-lesion-agnostic/monai/config_test.yml --data-split validation
``` 

The results of these inferences are stored in the same folder as the best model.

**Comment by plbenveniste** [comment]
I ran the following command line to get the performance of the model (for test set): 
```console
python ms-lesion-agnostic/monai/plot_performance.py --pred-dir-path ~/net/ms-lesion-agnostic/results/2024-07-18_10\:46\:21.634514/test_set/ --data-json-path ~/net/ms-lesion-agnostic/msd_data/dataset_2024-07-22_seed42_lesionOnly.json --split test
````

- On the test set

![dice_scores_contrast](https://github.com/user-attachments/assets/2e8aaa1d-2b2e-4f84-a364-a90d264c4a9d)
![dice_scores_orientation](https://github.com/user-attachments/assets/90e5d6f2-53fe-46a7-9481-0f03a1cf3b7b)
![dice_scores_site](https://github.com/user-attachments/assets/b5d98eba-2986-4f5c-a14f-980ceb48475a)

- On the validation set: 
- 
![dice_scores_contrast](https://github.com/user-attachments/assets/10ebb73e-86b0-481c-bc9b-85e7c4821d8b)
![dice_scores_orientation](https://github.com/user-attachments/assets/fcbbdf84-904c-49ea-976e-2b6a6bbc1640)
![dice_scores_site](https://github.com/user-attachments/assets/69250e66-6366-4f70-8318-46c75f46c882)

(Inference is still running on the train set)

**Comment by plbenveniste** [comment]
From a visual analysis of the validation set, no clear behaviour appeared: 
- bavaria-quebec-spine-ms-unstitched/sub-m058483/ses-20210310/anat/sub-m058483_ses-20210310_acq-ax_chunk-1_T2w.nii.gz 
One lesion missed
- basel-mp2rage/sub-P103/anat/sub-P103_UNIT1.nii.gz : 
Nothing particular
- ms-lesion-agnostic/data/bavaria-quebec-spine-ms-unstitched/sub-m333631/ses-20210525/anat/sub-m333631_ses-20210525_acq-ax_chunk-3_T2w.nii.gz 
One lesion missed
- canproco/sub-cal104/ses-M0/anat/sub-cal104_ses-M0_STIR.nii.gz
Small lesions missed
- canproco/sub-mon171/ses-M0/anat/sub-mon171_ses-M0_PSIR.nii.gz
Medium lesion missed in middle and big lesion missed at bottom
- nih-ms-mp2rage/sub-nih073/anat/sub-nih073_UNIT1.nii.gz
Missed pretty obvious lesions and segmented one in the middle of the brain
- sct-testing-large/sub-bwh001/anat/sub-bwh001_acq-sagstir_T2w.nii.gz
Segmented FP very small lesions, missed big lesion near brain stem
- sct-testing-large/sub-karo2011/anat/sub-karo2011_acq-sagcerv_T2star.nii.gz
Big FOV (head to thorathic): segmented a mini lesion near the breast
- sct-testing-large/sub-lyonOfsep022/anat/sub-lyonOfsep022_acq-sag_T2w.nii.gz
Missed lesion at top and segmented lesions in the middle (not sure its supposed to be here)
- sct-testing-large/sub-vanderbiltSeth013/anat/sub-vanderbiltSeth013_T2star.nii.gz
Segmented nothing eventhough a few obvious lesions

Moving on to a more precise analysis of the results.

**Comment by plbenveniste** [comment]
When looking at the results on the training set, I noticed that the model performs rather poorly on the basel dataset. I was wondering if maybe it was linked to the field of views of the images. 

On the train set: 
- sub-P011_UNIT1.nii.gz: image of the brain with lesion in upper spinal cord: pred misses pretty obvious lesions
- sub-P013_UNIT1.nii.gz: image of the brain with lesions in upper spinal cord: pred misses obivous lesions and segments a tiny one in the brain
- sub-P016_UNIT1.nii.gz: image of the brain with lesion in upper spinal cord: pred misses pretty obvious lesions
- sub-nih021_UNIT1.nii.gz: image of the brain with lesion in upper spinal cord: pred segments stuff in the brain
- sub-nih060_UNIT1.nii.gz: image of the brain with lesion in upper spinal cord: pred segments stuff in the brain
- sub-bwh026_acq-sagstir_T2w.nii.gz: image of cervical spinal cord: model misses pretty obvious lesion close to the brain
- sub-karo1898_acq-sagcerv_T2star.nii.gz: image cerv-thor of spinal cord: model misses pretty obvious lesions near C2-C3
- sub-lyonOfsep004_acq-sag_T2w.nii.gz: image of cerv-thor of spinal cord: model misses pretty obvious lesions near C1-C2

From this, I have two conclusions:
- it seems that the model doesn't perform well on images which contain a large portion of the brain
- it seems that the model often misses lesions around level C1 to C3 or in the brain stem

Next steps to be discussed.

**Comment by plbenveniste** [comment]
To deal with the problem caused by the differences in terms of protocol when segmenting lesions (some segment lesions in the brain stem and others stop at C1), we decided to remove the regions of the images which are above C1. To do so, we will use the contrast-agnostic model to segment the spinal cord and remove all voxels above the highest segmented point in the Inferior-superior direction.

**Comment by plbenveniste** [comment]
To reduce the variability of the AttentionUnet model, I tried doing some test-time-augmentation. 
The experiment was done using the script `monai/test_model_tta.py`.
```console
conda activate venv_monai
CUDA_VISIBLE_DEVICES=1 python ms-lesion-agnostic/monai/test_model_tta.py --config ms-lesion-agnostic/monai/config_test.yml --data-split test
```
We chose to perform 10 predictions for each subject which were stored in separate files. 
The dice scores were averaged using the following code: 
```console
python ms-lesion-agnostic/monai/average_tta_performance.py --pred-dir-path ~/moneta/users/pierrelouis/ms-lesion-agnostic/tta_exp/test_set
```
Then we plot the performance of the average dice: 
```console
python ms-lesion-agnostic/monai/plot_performance.py --pred-dir-path ~/net/ms-lesion-agnostic/tta_exp/test_set/ --data-json-path ~/net/ms-lesion-agnostic/msd_data/dataset_2024-07-24_seed42_lesionOnly.json --split test
```

Here is the output of this work with TTA: 
![dice_scores_contrast](https://github.com/user-attachments/assets/93965a56-eacb-4011-bb09-6395affeecca)

Here is a GIF showing the comparison between the performance without TTA and the performance with TTA. 
![ezgif com-animated-gif-maker](https://github.com/user-attachments/assets/c167e9d0-c99b-4b49-a580-6fc861981a25)

- [ ] I am now looking into examples of the algorithm having very different predictions for the same subject

**Comment by plbenveniste** [comment]
> [!WARNING]
> I just realize I was doing a big mistake in the test inference script. I am cropping the image to the size of the kernel: [64, 128, 128] in RPI before inference. Because of this, the model doesn't see the entire image. I am currently working on fixing this. This means I will also have to recompute the models performances

**Comment by plbenveniste** [comment]
I removed the cropping of the image before doing inference. It is now giving more accurate results. This can be seen with the fact that fewer predictions have a dice score of 1: this happened when the images were cropped and the cropped image didn't contain any lesion. 

However, it also raised the question of which threshold to use on the soft prediction: this is being explored in this issue https://github.com/ivadomed/ms-lesion-agnostic/issues/32

**Comment by plbenveniste** [comment]
Here are the result with TTA: 
![dice_scores_contrast](https://github.com/user-attachments/assets/eb6a8168-8eba-4807-aef5-f456a064b13e)

The result is disappointing. I think the strategy I used is problematic. What I was doing was creating multiple predictions, threshold at 0.5, then calculate each dice score and then average the dice score. 

TODO: 

- [ ] A strategy which would make more sense would be to create multiple predictions, threshold at 0.5, sum the predictions, threshold at 0.5 and then calculate the dice score.

**Comment by plbenveniste** [comment]
Here is the output without TTA: 
![dice_scores_contrast](https://github.com/user-attachments/assets/077cab56-f3c2-4bab-8cb0-801ab2f468a7)

The variability is much higher which shows that TTA does help reduce variability. But the average is higher as well. That confirms my hypothesis that when I do TTA, the results are lower because the average drops because out of 10 inference one is completely failed (dice score close to 0%).

**Comment by plbenveniste** [comment]
**TTA second strategy**

In a second attempt to improve the results of TTA, I did the following: 
perform inference 10 times with different data augmentations
threshold each predictions at 0.5
sum the predictions
threshold the summed prediction at 0.5 
computed the dice score

```console
python ms-lesion-agnostic/monai/compute_performance_tta_sum.py --path-pred ~/net/ms-lesion-agnostic/tta_exp/test_set/ --path-json ~/net/ms-lesion-agnostic/msd_data/dataset_2024-07-24_seed42_lesionOnly.json --split test --output-dir ~/net/ms-lesion-agnostic/tta_exp/perf_output_sum_image
```
Then to compute the figure, I did the following:

```console
python ms-lesion-agnostic/monai/plot_performance.py --pred-dir-path ~/net/ms-lesion-agnostic/tta_exp/perf_output_sum_image/ --data-json-path ~/net/ms-lesion-agnostic/msd_data/dataset_2024-07-24_seed42_lesionOnly.json --split test
```

Here is the output: 

![dice_scores_contrast](https://github.com/user-attachments/assets/aebde5f1-3736-46b7-a055-32f7e423b074)

**Comment by plbenveniste** [comment]
Now I want to evaluate the normal (meaning no TTA) AttentionUnet model on the following metrics: Dice, PPV, Sensitivity and F1 score. 
To do so, I modified the code in `monai/test_model.py` to include the computation of the other metrics.
I ran it with the following code: 
```console
CUDA_VISIBLE_DEVICES=1 python ms-lesion-agnostic/monai/test_model.py --config ms-lesion-agnostic/monai/config_test.yml --data-split test
```
To plot the performance: 
```console
python ms-lesion-agnostic/monai/plot_performance.py --pred-dir-path ~/net/ms-lesion-agnostic/results/2024-07-18_10\:46\:21.634514/test_set/ --data-json-path ~/net/ms-lesion-agnostic/msd_data/dataset_2024-07-22_seed42_lesionOnly.json --split test
```

Here are the results:

**Comment by plbenveniste** [comment]
Closing this issue as these results are outdated
